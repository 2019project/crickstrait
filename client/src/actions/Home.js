import { ALL_NEWS, NEWS_BY_ID } from "./Types";
import axios from "axios";
const url = "http://localhost:2001/api/home/";

// get all news - piyush
export const allNews = () => dispatch => {
  return axios
    .post(url + "allnews")
    .then(res => {
      dispatch({
        type: ALL_NEWS,
        payload: res.data.data
      });
    })
    .catch(err => {
      console.log(err);
    });
};

// get all news - piyush
export const newsById = news_id => dispatch => {
  return axios
    .post(url + news_id)
    .then(res => {
      dispatch({
        type: NEWS_BY_ID,
        payload: res.data.data
      });
    })
    .catch(err => {
      console.log(err);
    });
};
