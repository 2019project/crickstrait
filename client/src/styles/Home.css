@import url(https://fonts.googleapis.com/css?family=Karla);

* {
  font-family: Karla;
  user-select: none;
}

::-webkit-scrollbar {
  display: none;
}

#homecontainer {
  width: 97%;
  margin: 25px auto;
  display: grid;
  grid-template-columns: 1fr repeat(2, 1fr) 1fr;
  grid-gap: 30px;
}

#homeleftsidecontainer {
  /* grid-area: news; */
  grid-column: 1/2;
  height: 540px;
  border-radius: 4px;
  margin: 10px 0px 0px 0px;
  overflow-x: hidden;
  padding: 0px 10px 10px 10px;
  background-color: #fff;
  box-shadow: 0px 4px 6px 2px #cacaca;
  overflow-y: scroll;
}

#homeleftsidecontainer h1 {
  text-align: center;
  color: #2c3e50;
  padding: 0px 0px 10px 0px;
  margin: 0px;
  font-size: 30px;
}

#newsdivider hr {
  color: #2c3e50;
  border-top: 1px solid #2c3e50;
}

#homemiddlecontainer {
  grid-column: 2/4;
  height: 540px;
  padding: 0px 20px 20px 20px;
  overflow-x: hidden;
  background-color: #fff;
  /* box-shadow: 0px 4px 6px 2px #cacaca; */
  overflow-y: scroll;
}

#homerightsidecontainer {
  grid-column: 4/-1;
  height: 540px;
  overflow-x: hidden;
  font-size: 16px;
  background-color: #fff;
  box-shadow: 0px 4px 6px 2px #cacaca;
  margin: 10px 0px 0px 0px;
  border-radius: 4px;
  overflow-y: scroll;
}
#image {
  width: 100%;
  /* height: 100%; */
  padding: 0px 0px 10px 0px;
}

#recent-news-title {
  cursor: pointer;
  padding: 0px;
}

#homerightsidecontainer h1 {
  text-align: center;
  color: #2c3e50;
  padding: 0px;
  margin: 0px;
  font-size: 30px;
}

#indicesdivider hr {
  color: #fafafa;
  border-top: 1px solid #2c3e50;
}
#recent-news-title:hover {
  color: #39abf7;
}

#positiveIndex {
  color: #27ae60;
}

#negativeIndex {
  color: #ff4d4d;
}

#homeIndicesTable {
  border-collapse: collapse;
  border-spacing: 0;
  text-align: center;
  width: 100%;
  margin: auto;
}

#homeIndicesTable td {
  padding: 15px;
}

#homeIndicesTable tr:nth-child(odd) {
  background-color: #fafafa;
}

#homeIndicesTable th {
  text-align: left;
}

#homeIndicesTable tr:nth-child(1) {
  font-weight: bold;
  background: #fafafa;
}

#indicesName {
  cursor: pointer;
}

#indicesName:hover {
  color: #39abf7;
}

#headline {
  color: #2c3e50;
  font-family: karla;
  font-weight: 600;
  font-size: 24px;
  margin: 0px;
  padding: 0px 0px 10px 0px;
}

/* ------------------------------------------------------------------------ */

@media only screen and (max-width: 426px) {
  #homecontainer {
    width: 90%;
    grid-template-columns: 1fr;
  }
  #homeleftsidecontainer {
    grid-column: 1/4;
    grid-area: 3/1;
    height: 150px;
  }
  #homemiddlecontainer {
    grid-column: 1/4;
    grid-area: 2/1;
    height: 400px;
  }
  #homerightsidecontainer {
    grid-column: 1/4;
    grid-area: 1/1;
    height: 510px;
  }
}
