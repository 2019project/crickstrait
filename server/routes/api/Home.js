const express = require("express");
const router = express.Router();
const url = "mongodb://localhost:27017/";
var MongoClient = require("mongodb").MongoClient;

//geting all news from database - Piyush kumar
router.get("/allnews", async (req, res) => {
  MongoClient.connect(url, function(err, db) {
    if (err) throw err;
    var dbo = db.db("stocks");
    dbo
      .collection("news")
      .find({})
      .toArray(function(err, result) {
        if (err) throw err;
        // console.log(result);
        res.status(200).json({
          status: 200,
          data: result,
          message: "Retrieved all news Successfully"
        });
        db.close();
      });
  });
});

//geting news details with an id - Piyush kumar
router.get("/singleNews/:id", async (req, res, next) => {
  MongoClient.connect(url, function(err, db) {
    try {
      let id = req.params.id;
      var dbo = db.db("stocks");
      dbo.collection("news").findOne({ new_id: +id }, function(err, result) {
        if (!result) {
          return res.status(400).send({ message: "News not found" });
        } else {
          if (err) throw err;

          //   console.log(result);
          res.status(200).json({
            status: 200,
            data: result,
            message: "Retrieved news Successfully"
          });
          db.close();
        }
      });
    } catch (err) {
      next(err);
    }
  });
});

router.get("/index", async (req, res) => {
  console.log("INdices");
  MongoClient.connect(url, function(err, db) {
    if (err) throw err;
    var dbo = db.db("stocks");
    dbo
      .collection("stocks_data")
      .find({ isIndex: true })
      .toArray(function(err, result) {
        console.log(result);
        if (err) throw err;
        // console.log(result);
        res.status(200).json({
          status: 200,
          data: result,
          message: "Retrieved all news Successfully"
        });
        db.close();
      });
  });
});

module.exports = router;
