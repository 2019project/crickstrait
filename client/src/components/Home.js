import React, { Component } from "react";
import { allNews, newsById, getIndices } from "../actions/Home";
import { connect } from "react-redux";
import "../styles/Home.css";
import welcome from "./welcome.svg";
import "react-loader-spinner/dist/loader/css/react-spinner-loader.css";
import loader from "./Common/Loader.gif";
import Loader from "react-loader-spinner";

export class Home extends Component {
  componentDidMount() {
    //this function calls the get News function which is defined in the actions which will retrive all the data from the news table.
    this.props.allNews();
    this.props.newsById(1);
    this.props.getIndices();
  }

  render() {
    console.log(this.props.singleNews);
    console.log(this.props.isLoading);
    return (
      <body>
        <div id="homecontainer">
          {/* <Navbar> */}
          {this.props.isLoading ? ( //use to display loader [piyush]
            <div style={{ margin: "200px 500px" }}>
              <Loader type={Loader} color="#2c3e50" height="100" width="400" />
              {/* <img src={loader} alt="loading..." /> */}
            </div>
          ) : (
            <>
              <div id="homeleftsidecontainer">
                <h1>News</h1>
                <div>
                  <div id="recent-news"></div>
                  {/* recent news */}
                  {this.props.news.map((news, index) => (
                    <div className="div-newspage">
                      <div id="news-list">
                        <p
                          id="recent-news-title"
                          onClick={() => this.props.newsById(news.new_id)}
                        >
                          {news.headline}
                        </p>
                        <p id="newsdivider">
                          <hr />
                        </p>
                      </div>
                    </div>
                  ))}
                </div>
              </div>
              <div id="homemiddlecontainer">
                {/* <div>
                  <h3 id="headline">{this.props.singleNews[0].headline}</h3>
                  <img
                    id="image"
                    src={
                      "data:image/jpeg;base64," +
                      this.props.singleNews[0].news_image
                    }
                  />
                  <p id="headlineDescription">
                    {this.props.singleNews[0].description}
                  </p>
                </div> */}
              </div>
              <div id="homerightsidecontainer">
                <h1>Indices</h1>
                <div>
                  <table id="homeIndicesTable">
                    <th>Indices</th>
                    <th>Last</th>
                    <th>%Chng</th>
                    {this.props.indices.map((indices, index) => (
                      <tr>
                        <td
                          id="indicesName"
                          onClick={() =>
                            this.props.history.push(
                              "/indexProfile/" + indices.ticker_id,
                              {
                                indices
                              }
                            )
                          }
                        >
                          {indices.ticker_name}
                        </td>
                        <td>
                          {Number(
                            indices.ticker_dates["2019-11-05"].closing
                          ).toFixed(2)}
                        </td>
                        <td
                          id={
                            String(
                              (indices.ticker_dates["2019-10-31"].closing -
                                indices.ticker_dates["2019-11-05"].closing) /
                                100
                            ).charAt(0) == "-"
                              ? "negativeIndex"
                              : "positiveIndex"
                          }
                        >
                          {Number(
                            (indices.ticker_dates["2019-10-31"].closing -
                              indices.ticker_dates["2019-11-05"].closing) /
                              100
                          ).toFixed(3)}
                        </td>
                      </tr>
                    ))}{" "}
                  </table>
                </div>
              </div>
            </>
          )}
        </div>
      </body>
    );
  }
}

const mapStateToProps = state => ({
  news: state.homeReducer.news,
  singleNews: state.homeReducer.singleNews,
  indices: state.homeReducer.indices,
  isLoading: state.LoadingReducer.isLoading
});
export default connect(
  mapStateToProps,
  {
    allNews,
    newsById,
    getIndices
  }
)(Home);
