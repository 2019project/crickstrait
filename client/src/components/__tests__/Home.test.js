import React from "react";
import { Home } from "../Home";
import { shallow, mount } from "enzyme";

const home = jest.fn();
const allNews = jest.fn();
const newsById = jest.fn();
const getIndices = jest.fn();
const news = [
  { new_id: 1, headline: "Master Blaster Sachin baby" },
  { new_id: 2, headline: "BCCI President elections coming up" }
];
const singleNews = [];
const indices = [];
const wrapper = mount(
  <Home
    home={home}
    news={news}
    singleNews={singleNews}
    allNews={allNews}
    newsById={newsById}
    getIndices={getIndices}
    indices={indices}
  />
);

describe("Testing home Component", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it("should mount the component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should check for headline description in news section", () => {
    expect(wrapper.find("#recent-news-title").text()).toBe(news[0].headline);
  });

  it("should simulate  News Click", () => {
    expect(
      wrapper
        .find("#recent-news-title")
        .at(0)
        .simulate("click")
    );
  });
  it("checks for componenetDidMount  to be called", () => {
    const componentDidMount = jest.spyOn(Home.prototype, "componentDidMount");
    wrapper.instance().componentDidMount();
    expect(componentDidMount).toHaveBeenCalled();
  });
});
