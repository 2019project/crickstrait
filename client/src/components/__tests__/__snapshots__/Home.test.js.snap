// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing home Component should mount the component 1`] = `
<Home
  allNews={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  getIndices={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  home={[MockFunction]}
  indices={Array []}
  news={
    Array [
      Object {
        "headline": "Master Blaster Sachin baby",
        "new_id": 1,
      },
      Object {
        "headline": "BCCI President elections coming up",
        "new_id": 2,
      },
    ]
  }
  newsById={
    [MockFunction] {
      "calls": Array [
        Array [
          1,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  singleNews={Array []}
>
  <body>
    <div
      id="homecontainer"
    >
      <div
        id="homeleftsidecontainer"
      >
        <h1>
          News
        </h1>
        <div>
          <div
            id="recent-news"
          />
          <div
            className="div-newspage"
          >
            <div
              id="news-list"
            >
              <p
                id="recent-news-title0"
                onClick={[Function]}
              >
                Master Blaster Sachin baby
              </p>
              <p
                id="newsdivider"
              >
                <hr />
              </p>
            </div>
          </div>
          <div
            className="div-newspage"
          >
            <div
              id="news-list"
            >
              <p
                id="recent-news-title1"
                onClick={[Function]}
              >
                BCCI President elections coming up
              </p>
              <p
                id="newsdivider"
              >
                <hr />
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        id="homemiddlecontainer"
      >
        <div>
          <h3
            id="headline"
          />
          <img
            id="image"
            src="welcome.svg"
          />
          <p
            id="headlineDescription"
          />
        </div>
      </div>
      <div
        id="homerightsidecontainer"
      >
        <h1>
          Indices
        </h1>
        <div>
          <table
            id="homeIndicesTable"
          >
            <th>
              Indices
            </th>
            <th>
              Last
            </th>
            <th>
              Chng %
            </th>
             
          </table>
        </div>
      </div>
    </div>
  </body>
</Home>
`;
